File: src/main/java/org/mtransit/android/ad/IAdManager.java
package org.mtransit.android.ad;

import android.content.res.Configuration;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import org.mtransit.android.ui.view.common.IActivity;

public interface IAdManager {

	void init(@NonNull IActivity activity);

	void onHasAgenciesEnabledUpdated(@NonNull IActivity activity, @Nullable Boolean hasAgenciesEnabled);

	void setShowingAds(@Nullable Boolean newShowingAds, @NonNull IActivity activity);

	int getRewardedAdAmount();

	long getRewardedAdAmountInMs();

	void linkRewardedAd(@NonNull IActivity activity);

	void unlinkRewardedAd(@NonNull IActivity activity);

	void refreshRewardedAdStatus(@NonNull IActivity activity);

	boolean isRewardedAdAvailableToShow();

	@SuppressWarnings("UnusedReturnValue")
	boolean showRewardedAd(@NonNull IActivity activity);

	int getBannerHeightInPx(@Nullable IActivity activity);

	void adaptToScreenSize(@NonNull IActivity activity, @Nullable Configuration configuration);

	void resumeAd(@NonNull IActivity activity);

	void pauseAd(@NonNull IActivity activity);

	void destroyAd(@NonNull IActivity activity);

	long getRewardedUntilInMs();

	void resetRewarded();

	boolean isRewardedNow();

	void setRewardedAdListener(@Nullable RewardedAdListener rewardedAdListener);

	void openAdInspector();

	boolean shouldSkipRewardedAd();

	interface RewardedAdListener {
		void onRewardedAdStatusChanged();

		boolean skipRewardedAd();
	}
}

File: app-android/src/main/java/org/mtransit/android/ad/IAdManager.kt
package org.mtransit.android.ad

import android.content.res.Configuration
import androidx.annotation.MainThread
import androidx.annotation.WorkerThread
import androidx.lifecycle.LiveData
import org.mtransit.android.ui.view.common.IActivity
import kotlin.time.Instant

interface IAdManager {

    fun init(activity: IAdScreenActivity, onInitCompleteListener: () -> Unit = {}, withConsentOnly: Boolean = false)
    fun initForScreens(activity: IAdScreenActivity)

    suspend fun onHasAgenciesEnabledUpdated(hasAgenciesEnabled: Boolean?, activity: IAdScreenActivity)

    suspend fun initHasSubscriptionFromCache()

    fun canShowAds(): Boolean?

    suspend fun setHasSubscription(hasSubscription: Boolean?, activity: IAdScreenActivity)

    // region App open ad

    fun isAppOpenAdAvailable(): Boolean

    fun isShowingAppOpenAd(): Boolean

    @MainThread
    fun loadAppOpenAd(): Boolean

    fun showAppOpenAdIfAvailable(activity: IActivity, onShowAdComplete: () -> Unit)

    // endregion App open ad

    // region Rewarded ad

    val rewardedAdAmountInDays: Int

    fun linkRewardedAd(activity: IActivity)

    fun unlinkRewardedAd(activity: IActivity)

    suspend fun refreshRewardedAdStatus(activity: IActivity)

    fun isRewardedAdAvailableToShow(): Boolean

    fun showRewardedAd(activity: IActivity): Boolean

    // endregion Rewarded ad

    fun getBannerHeightInPx(activity: IAdScreenActivity?): Int

    fun adaptToScreenSize(activity: IAdScreenActivity, configuration: Configuration? = activity.activity?.resources?.configuration)

    fun onResumeScreen(activity: IAdScreenActivity)

    fun onTimeChanged(activity: IAdScreenActivity)

    @Suppress("unused")
    fun resumeAd(activity: IAdScreenActivity)

    fun pauseAd(activity: IAdScreenActivity)

    fun destroyAd(activity: IAdScreenActivity)

    val rewardedUntilLive: LiveData<Instant>
    val rewardedNowLive: LiveData<Boolean>

    @WorkerThread
    fun getRewardedUntil(): Instant

    fun resetRewarded()

    @WorkerThread
    fun isRewardedNow(): Boolean

    fun setRewardedAdListener(rewardedAdListener: RewardedAdListener?)

    fun openAdInspector(activity: IActivity)

    @WorkerThread
    fun shouldSkipLoadingRewardedAd(): Boolean

    interface RewardedAdListener {
        fun onRewardedAdStatusChanged()

        @WorkerThread
        fun skipLoadingRewardedAd(): Boolean
    }
}
